### DISCUSSION TITLE
can not find elements on the bottom of the page

### DISCUSSION THREAD

**Discussion by Shahin-rmz** [zauberzeug/nicegui#632] [Q&A]
Hi,
My NiceGUI app, gets a `pandas Dataframe` and gives it back as a `table`.
There is a problem with find in the browser.
User **has to** scroll down in order to find words on the bottom of the page.
I can not not understand it to know how to handle it.

Here is upper side of the page, the number of find elements are 0.
![image](https://user-images.githubusercontent.com/32805117/227737926-aab5d12f-ef22-4cae-9b3e-e7eb90b29b4a.png)

Here is when I have already scrolled down.
I can find it without problem.
![image](https://user-images.githubusercontent.com/32805117/227738038-68cd9c40-d67d-4450-be7b-cde4d3ae7196.png)

**Comment by falkoschindler** [comment]
Interesting observation, @Shahin-rmz!
It looks like AG Grid does quite some performance optimization on the client. Let's try a minimal example:

```py
grid = ui.aggrid({
    'columnDefs': [{'field': 'number'}],
    'rowData': [{'number': f'{i}'} for i in range(100)],
})
```

Initially the grid shows only 8 rows. If you use the element inspector of your browser and search for `div.ag-center-cols-container` you'll find it has 18 children. When you scroll down, the number of child elements increases, but gets never larger than 29. So it looks like AG Grid always holds 10 more rows in each scroll direction than currently (partly) visible. These elements are re-used and dynamically filled with new values. This saves resources and allows to display thousands of rows with ease. But it means that there _is_ no element "6609" on the page until you scroll down.

You'll probably find some more details on their website. This is only what I found out by playing around with the example. Anyway, it's an AG Grid feature/limitation. Maybe you can configure it somewhere. But your application would probably not be possible without this smart cell re-use.

**Reply by Shahin-rmz** [reply]
Thanks for the answer.
I will defenentely search in `ag-grid` website for further information.
But as of now, I have also tried out `ut.table` and `ui.aggrid` both.
it seems that both are `ag-grid` and `ui.table` is deprecated right?
Nicegui version : `nicegui==1.1.11`

**Reply by Shahin-rmz** [reply]
After some research, I found out `minifilter` is THE solution.
Here is the [page](https://ag-grid.com/javascript-data-grid/filter-set-mini-filter/#custom-searches)
It's pretty interesting though.
But I am not sure how to implement it in `nicegui`

```python
df = pd.read_csv('wc.csv')

data_dict = {
    'columnDefs': [
        {'headerName': col, 'field': col} for col in df.columns
    ],
    'rowData': df.to_dict('records'),
    'rowSelection': 'multiple',
}


print(data_dict)

ui.aggrid(data_dict).style('height: calc(100vh - 32px)')
ui.run()

```
here is the code.

**Reply by falkoschindler** [reply]
During the transition to a new naming scheme we deprecated `ui.table`. Since NiceGUI 1.2.0 it is now as follows:

- `ui.table` creates a Quasar table (like most NiceGUI elements are based on Quasar)
- `ui.aggrid` creates an AG Grid (named after a third-party dependency like Plotly or Mermaid)

---

I didn't know about minifilters. But they look very promising! This just works:

```py
grid = ui.aggrid({
    'columnDefs': [{'field': 'number', 'filter': 'agTextColumnFilter', 'floatingFilter': True}],
    'rowData': [{'number': f'{i}'} for i in range(100)],
})
```

And you can use other filters like `'agNumberColumnFilter'`.

**Reply by Shahin-rmz** [reply]
Thanks for the hint.
I didn't know that we should write all the parameters as `string`.
The code works like a charm

Here the user searched and found the last row.
![image](https://user-images.githubusercontent.com/32805117/227747048-8a5c1f0b-187d-45aa-ab1c-86307a9e40d3.png)

Here is how the code look like.
```python
data_dict = {
    'columnDefs': [
        {'headerName': col, 'field': col,
         'filter':'agSetColumnFilter',} for col in df.columns
    ],
    'rowData': df.to_dict('records'),
    'rowSelection': 'multiple',
}


print(data_dict)

ui.table(data_dict).style('height: calc(100vh - 32px)')
ui.run()

```

**Comment by rodja** [comment]
@Shahin-rmz that would be a very helpful live-example on the [aggrid documentation page](https://github.com/zauberzeug/nicegui/blob/main/website/more_documentation/aggrid_documentation.py). We would love to see a pull-request from you to enhance the docs. We recently [added a section in the CONTRIBUTING.md file](https://github.com/zauberzeug/nicegui/blob/main/CONTRIBUTING.md#additional-demos) which lists the few simple steps to do such enhancements.

**Comment by Shahin-rmz** [comment]
@rodja Thanks for comment.
Although I think this comment is regarding the other discussion that I recently participated.
BTW, I will surely `fork` , enhance the documentation and send a `pr` to the project.
I really like the approach of NiceGUI regarding the problem with web solutions with Python.
Zauberzeug is dedicating resources to develop and maintain this FOSS project. I even saw @falkoschindler sending `pr` on sunday.
All respect from my side to Zauberzeug team.
Best

**Comment by falkoschindler** [comment]
Sorry @Shahin-rmz, I somehow overlooked that you agreed to submit a PR. I just added the demo myself. 😕

**Reply by Shahin-rmz** [reply]
I have absolutely nothing to complain though :smile:
